Packet++/src/VrrpLayer.cpp

```diff
@@ -0,0 +1,517 @@
+			}
+			copyIPAddressToData(newIpAddresses + ipAddrOffset, ipAddress);
+
+			ipAddrOffset += ipAddrLen;
+		}
+
+		getPacketPtr()->ipAddrCount = getIPAddressesCount() + ipAddresses.size();
+
+		return true;
+	}
+
+	bool VrrpLayer::addIPAddresses(const std::vector<IPAddress> &ipAddresses) {
+		return addIPAddressesAt(ipAddresses, (int) getHeaderLen());
+	}
+
+	bool VrrpLayer::addIPAddress(IPAddress &ipAddress) {
+		std::vector<IPAddress> ipAddresses;
+		ipAddresses.push_back(ipAddress);
+
+		return addIPAddressesAt(ipAddresses, (int) getHeaderLen());
+	}
+
+	bool VrrpLayer::removeIPAddressAtIndex(int index) {
+		int ipAddressCount = (int) getIPAddressesCount();
+
+		if (index < 0 || index >= ipAddressCount) {
+			PCPP_LOG_ERROR("Cannot remove virtual IP address, index " << index << " is out of bounds");
+			return false;
+		}
+
+		size_t ipAddressLen = getIPAddressLen();
+		if (ipAddressLen == 0) {
+			PCPP_LOG_ERROR("Cannot remove IP address, for ip address length is invalid.");
+			return false;
+		}
+
+		size_t offset = VRRP_PACKET_FIX_LEN;
+		uint8_t *curIpAddress = getFirstIPAddress();
+		for (int i = 0; i < index; i++) {
+			if (curIpAddress == nullptr) {
+				PCPP_LOG_ERROR("Cannot remove virtual IP address at index "
+									   << index << ", cannot find virtual IP address at index " << i);
+				return false;
+			}
+
+			offset += ipAddressLen;
+			curIpAddress = getNextIPAddress(curIpAddress);
+		}
+
+		if (!shortenLayer((int) offset, ipAddressLen)) {
+			PCPP_LOG_ERROR("Cannot remove virtual IP address at index " << index << ", cannot shorted layer");
+			return false;
+		}
+
+		getPacketPtr()->ipAddrCount = ipAddressCount - 1;
+
+		return true;
+	}
+
+	bool VrrpLayer::removeAllIPAddresses() {
+		size_t offset = VRRP_PACKET_FIX_LEN;
+		size_t packetLen = getHeaderLen();
+		if (packetLen <= offset) {
+			return false;
+		}
+
+		if (!shortenLayer((int)offset, packetLen - offset)) {
+			PCPP_LOG_ERROR("Cannot remove all virtual IP address(es), cannot shorted layer");
+			return false;
+		}
+
+		getPacketPtr()->ipAddrCount = 0;
+
+		return true;
+	}
+
+	void VrrpLayer::copyIPAddressToData(uint8_t *data, const IPAddress &ipAddress) const {
+		size_t ipAddressLen = getIPAddressLen();
+		if (ipAddressLen == 0) {
+			PCPP_LOG_ERROR("Cannot copy virtual IP address to data, for ip address length is invalid.");
+			return;
+		}
+
+		if (ipAddress.isIPv4()) {
+			memcpy(data, ipAddress.getIPv4().toBytes(), ipAddressLen);
+		} else if (ipAddress.isIPv6()) {
+			memcpy(data, ipAddress.getIPv6().toBytes(), ipAddressLen);
+		} else {
+		}
+	}
+
+	bool VrrpLayer::getIPAddressFromData(uint8_t *data, IPAddress &ipAddress) const {
+		if (getAddressType() == IPAddress::IPv4AddressType) {
+			IPv4Address ipv4Address(*((uint32_t *) data));
+			ipAddress = ipv4Address;
+		} else if (getAddressType() == IPAddress::IPv6AddressType) {
+			IPv6Address ipv6Address(data);
+			ipAddress = ipv6Address;
+		} else {
+			return false;
+		}
+
+		return true;
+	}
```

Comment:
A nicer implementation for this method:
```cpp
IPAddress VrrpLayer::getIPAddressFromData(uint8_t *data) const {
	if (getAddressType() == IPAddress::IPv4AddressType) {
		return IPv4Address(*((uint32_t *) data));
	} else {
		return IPv6Address(data);
}
```

Reply:
Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/0233d684d8aff42029b747845745d835aa066beb